0000-zmix/README.md

```diff
@@ -0,0 +1,395 @@
+implementation of ZKPs which can be consumed through a simple API, making
+it easier for other projects to use ZKPs.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Introduction
+
+A zero-knowledge proof enables a prover to convince a verifier of the truth
+of some statement about certain values, without revealing those values. For
+example, a prover may know that values `A` and `B` have an equal discrete
+logarithm. That is, for some `x`, `A = g^x` and `B = h^x`. It could
+convince a verifier that this is true by revealing `x`. If the prover would
+rather not disclose `x`, it could instead construct a zero-knowledge proof
+of this statement, and send the resulting proof to the verifier. The 
+verifier can now be sure that `log_g(A) = log_h(B)`, without knowing the
+discrete log `x`.
+
+In the example above, there are three components: 
+
+- What do we want to prove? That there exists an `x` such that `A = g^x`
+and `B = h^x`.
+
+- How do we know this is true? Because we know `x`.
+
+- How can we prove it? With a zero-knowledge proof that the prover produces
+and the verifier verifies.
+
+In zmix, we have data types corresponding to these three components:
+
+- `ProofSpec`: specifies what we want to prove
+
+- `Witness`: contains the secrets that satisfy the `ProofSpec`.
+
+- `Proof`: the zero-knowledge proof. 
+
+The library provides two features: construction and verification of
+zero-knowledge proofs. 
+
+To construct a proof, zmix requires a `ProofSpec s` and a `Witness w`
+(which is a valid witness for `p`), and will output a `Proof p`. 
+
+To verify a proof, zmix requires a `ProofSpec s` and a `Proof p`, and will
+output a boolean indicating whether the proof was valid (true) or invalid
+(false).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Conceptually, the zmix library will offer the functions:
+
+* `generate_proof(s: ProofSpec, w: Witness) -> Result<Proof, Error>`
+* `verify_proof(p: Proof, s: ProofSpec) -> Result<(), Error>`
+
+where the proof specification contains
+
+1. the number of secrets involved in the zero-knowlege proof (we call these
+_messages_), and
+1. a list of _statements_, where each statement represents a _part_ of the overall zero-knowlege proof:
+
+```
+pub struct ProofSpec {
+    message_count: usize,
+    statements: Vec<Statement>,
+    ...
+}
+```
+The library will offer and support various types of statements, such as signatures and commitments.
+To combine and connect all the statements (i.e., the proof's parts) contained in the proof specification, that is, glue all parts together into a single zero-knowledge proof, the library will use _Schnorr proofs_.
+
+For generating a Schnorr proof that encompasses multiple statements, for each statement one requires a _hash contribution_ on the one hand, and a _proof contribution_ based on an aggregation of the hash contributions on the other hand.
```

Comment:
"a proof contribution based on an aggregation of the hash contributions on the other hand" is incomplete. The proof contribution also involves the witness.